server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -177,13 +183,52 @@ public Index selectTimeSeriesWriteIndex(Instant timestamp, Metadata metadata) {
         return null;
     }
 
-    public boolean isTimeSeries(Function<Index, IndexMetadata> indices) {
-        return isTimeSeries(indices.apply(getWriteIndex()));
-    }
-
-    public boolean isTimeSeries(IndexMetadata indexMetadata) {
-        IndexMode indexMode = IndexSettings.MODE.get(indexMetadata.getSettings());
-        return indexMode == IndexMode.TIME_SERIES;
+    /**
+     * Validates this data stream. If this is a time serie data stream then this method validates that temporal range
+     * of backing indices (defined by index.time_series.start_time and index.time_series.end_time) do not overlap with each other.
+     *
+     * @param imSupplier Function that supplies {@link IndexMetadata} instances based on the provided index name
+     */
+    public void validate(Function<String, IndexMetadata> imSupplier) {
+        if (indexMode == IndexMode.TIME_SERIES) {
+            // Get a sorted overview of each backing index with there start and end time range:
+            List<Tuple<String, Tuple<Instant, Instant>>> startAndEndTimes = indices.stream()
+                .map(index -> imSupplier.apply(index.getName()))
+                .map(im -> {
+                    Instant start = IndexSettings.TIME_SERIES_START_TIME.get(im.getSettings());
+                    Instant end = IndexSettings.TIME_SERIES_END_TIME.get(im.getSettings());
+                    assert end.isAfter(start); // This is also validated by TIME_SERIES_END_TIME setting.
+                    return new Tuple<>(im.getIndex().getName(), new Tuple<>(start, end));
+                })
+                .sorted(Comparator.comparing(entry -> entry.v2().v1())) // Sort by start time
+                .collect(Collectors.toList());
+
+            Tuple<String, Tuple<Instant, Instant>> previous = null;
+            for (Tuple<String, Tuple<Instant, Instant>> current : startAndEndTimes) {
+                if (previous == null) {
+                    previous = current;
+                } else {
+                    // The end_time of previous backing index should be equal or less than start_time of current backing index.
+                    // If previous.end_time > current.start_time then we should fail here:
+                    if (previous.v2().v2().compareTo(current.v2().v1()) > 0) {
+                        var formatter = DateFieldMapper.DEFAULT_DATE_TIME_FORMATTER;
```

Comment:
Is this var just used to be shorter than the static `DEFAULT_DATE_TIME_FORMATTER`? We probably don't need to assign it in the loop right? It's fine either way, since this is likely to get optimized out, I am just curious about it

Reply:
> Is this var just used to be shorter than the static DEFAULT_DATE_TIME_FORMATTER? 

Yes.

> We probably don't need to assign it in the loop right?

Yes and I moved it out of the loop.